modules/contact-form/js/gutenblock.jsx

```diff
@@ -0,0 +1,568 @@
+					/>
+				</PanelBody>
+			</InspectorControls>
+		);
+	}
+}
+
+class GrunionFieldLabel extends Component {
+	render() {
+		return (
+			<Fragment>
+				<input
+					type='text'
+					value={ this.props.label }
+					className='grunion-field-label'
+					onChange={ ( x )=>this.props.setAttributes( { label: x.target.value } ) }
+				/>
+				{ this.props.required && <span className="required">{ __( '(required)' ) }</span> }
+			</Fragment>
+		);
+	}
+}
+
+class GrunionField extends Component {
+	render() {
+		return (
+			<Fragment>
+				<GrunionFieldSettings
+					required={ this.props.required }
+					setAttributes={ this.props.setAttributes }
+				/>
+				<div className="grunion-field">
+					<TextControl
+						type={ this.props.type }
+						label={ <GrunionFieldLabel
+							required={ this.props.required }
+							label={ this.props.label }
+							setAttributes={ this.props.setAttributes }
+						/> }
+						disabled={ true }
+					/>
+				</div>
+			</Fragment>
+		);
+	}
+}
+
+class GrunionFieldTextarea extends Component {
+	render() {
+		return (
+			<Fragment>
+				<GrunionFieldSettings
+					required={ this.props.required }
+					setAttributes={ this.props.setAttributes }
+				/>
+				<div className="grunion-field">
+					<TextareaControl
+						label={ <GrunionFieldLabel
+							required={ this.props.required }
+							label={ this.props.label }
+							setAttributes={ this.props.setAttributes }
+						/> }
+						disabled={ true }
+					/>
+				</div>
+			</Fragment>
+		);
+	}
+}
+
+class GrunionFieldCheckbox extends Component {
+	render() {
+		return (
+			<Fragment>
+				<GrunionFieldSettings
+					required={ this.props.required }
+					setAttributes={ this.props.setAttributes }
+				/>
+				<div className="grunion-field">
+					<CheckboxControl
+						label={ <GrunionFieldLabel
+							required={ this.props.required }
+							label={ this.props.label }
+							setAttributes={ this.props.setAttributes }
+						/> }
+						disabled={ true }
+					/>
+				</div>
+			</Fragment>
+		);
+	}
+}
+
+class GrunionFieldMultiple extends Component {
+	render() {
+		return (
+			<Fragment>
+				<GrunionFieldSettings
+					required={ this.props.required }
+					setAttributes={ this.props.setAttributes }
+				/>
+				<div className="grunion-field">
+					<GrunionFieldLabel
+						required={ this.props.required }
+						label={ this.props.label }
+						setAttributes={ this.props.setAttributes }
+					/>
+					<ol>
+						{ _.map( this.props.options, ( option, key, list )=>(
+							<li key={ key }>
+								<input
+									type="text"
+									className="option"
+									value={ option }
+									placeholder={ __( 'Enter your option value here…' ) }
+									onChange={ function( x ) {
+										const $options = jQuery( x.target ).closest( 'ol' ).find( 'input.option' );
+										this.props.setAttributes( {
+											options: _.pluck( $options.toArray(), 'value' )
+										} );
+									}.bind( this ) }
```

Comment:
I've also gotten some warnings about the `.bind()` here -- any recommendations on how to rewrite it instead?

Reply:
the bind here is fine. The warnings are related to possible perf problems with passing a new function instance to the component on each render, which can cause the component to re-render unnecessarily.

If you really want to rewrite it, I'd do two things: Make up a new GrunionOption component that takes the option and an onChange callback as props. Then in the GrunionOption component, it would register the onChange handler against a locally bound function and call this.props.onChange, passing out the option that was passed in.  The option component would look something like:
```
class GrunionOption extends React.Component {
  handleOnChange = event => {
    this.props.onChange( this.props.option )
  }

  render() {
    ...
    return <input onChange={ this.handleChange } ... />
  }
}

Reply:
Same reasoning as above. The creation of a new function each time this renders causes the subcomponent to no longer match its props and forces it to re-render to. The bind does the same. I'd make an `onChange` function outside of render and pre-bind it in the constructor (or use arrow function and it binds automatically.)